Yosemite/Yosemite/Stores/ReceiptStore.swift

```diff
@@ -69,31 +87,52 @@ private extension ReceiptStore {
     }
 
     func generateCartTotals(order: Order, parameters: CardPresentReceiptParameters) -> [ReceiptTotalLine] {
-        var totalLines = [ReceiptTotalLine]()
-        if NSDecimalNumber(apiAmount: order.totalTax).decimalValue > 0.00 {
-            totalLines.append(ReceiptTotalLine(description: ReceiptContent.Localization.totalTaxLineDescription,
-                                          amount: order.totalTax))
-        }
-        totalLines.append(ReceiptTotalLine(description: ReceiptContent.Localization.amountPaidLineDescription, amount: parameters.formattedAmount))
-        return totalLines
+        let subtotalLines = [discountLine(order: order),
+                             lineIfNonZero(description: ReceiptContent.Localization.shippingLineDescription,
+                                           amount: order.shippingTotal),
+                             lineIfNonZero(description: ReceiptContent.Localization.totalTaxLineDescription,
+                                           amount: order.totalTax)]
+            .compactMap { $0 }
+        let totalLine = [ReceiptTotalLine(description: ReceiptContent.Localization.amountPaidLineDescription,
+                                         amount: parameters.formattedAmount)]
+
+        return subtotalLines + totalLine
     }
 
-    func generateReceiptContent(order: Order, parameters: CardPresentReceiptParameters) -> ReceiptContent {
-        let lineItems = generateLineItems(order: order)
-        let cartTotals = generateCartTotals(order: order, parameters: parameters)
+    func discountLine(order: Order) -> ReceiptTotalLine? {
+        let discountValue = NSDecimalNumber(apiAmount: order.discountTotal).decimalValue
+        if discountValue == 0 && order.coupons.isEmpty {
```

Comment:
I keep wondering if we want to check both these conditions. In the most common case, if there is a discount it will be because a coupon was applied. However, there are ways that these two things could go in separate ways:
- You can create a coupon with $0 value but that [allows free shipping](https://docs.woocommerce.com/document/free-shipping/#section-3)
- I see several extensions that let you set up discounts when you bundle products, or are quantity based, which would show up as a discount with no coupon.

I think in this second case we'd still want to show the Discount line without a coupon code, so the prices add up?

Reply:
I agree, and with the existing code both of these examples will show up... the only time this line won't show is for a Zero discount where no coupons were applied.